source/common/common/compiled_string_map.h

```diff
@@ -0,0 +1,208 @@
+    while (range_start != initial.end()) {
+      // Find the first key whose length differs from the current key length.
+      // Everything in between is keys with the same length.
+      auto range_end =
+          std::find_if(range_start, initial.end(), [len = range_start->first.size()](const KV& e) {
+            return e.first.size() != len;
+          });
+      std::vector<KV> node_contents;
+      // Populate a FindFn for the nodes in that range.
+      node_contents.reserve(range_end - range_start);
+      std::copy(range_start, range_end, std::back_inserter(node_contents));
+      table_[range_start->first.size()] = createEqualLengthNode(node_contents);
+      range_start = range_end;
+    }
+  }
+
+private:
+  /*
+   * @param node_contents the set of key-value pairs that will be children of
+   *                      this node.
+   */
+  static std::unique_ptr<Node> createEqualLengthNode(std::vector<KV> node_contents) {
+    if (node_contents.size() == 1) {
+      return std::make_unique<LeafNode>(node_contents[0].first, std::move(node_contents[0].second));
+    }
+    struct IndexSplitInfo {
+      uint8_t index, min, max, count;
+    } best{0, 0, 0, 0};
+    for (size_t i = 0; i < node_contents[0].first.size(); i++) {
+      std::array<bool, 256> hits{};
+      IndexSplitInfo info{static_cast<uint8_t>(i), 255, 0, 0};
+      for (size_t j = 0; j < node_contents.size(); j++) {
+        uint8_t v = node_contents[j].first[i];
+        if (!hits[v]) {
+          hits[v] = true;
+          info.count++;
+          info.min = std::min(v, info.min);
+          info.max = std::max(v, info.max);
+        }
+      }
+      if (info.count > best.count) {
+        best = info;
+      }
+    }
+    std::vector<std::unique_ptr<Node>> nodes;
+    nodes.resize(best.max - best.min + 1);
+    std::sort(node_contents.begin(), node_contents.end(), [&best](const KV& a, const KV& b) {
+      return a.first[best.index] < b.first[best.index];
+    });
+    auto range_start = node_contents.begin();
+    // Populate the sub-nodes for each character-branch.
+    while (range_start != node_contents.end()) {
+      // Find the first key whose character at position [best.index] differs from the
+      // character of the current range.
+      // Everything in between is keys with the same character at this index.
+      auto range_end = std::find_if(range_start, node_contents.end(),
+                                    [index = best.index, c = range_start->first[best.index]](
+                                        const KV& e) { return e.first[index] != c; });
+      // Possible optimization was tried here, std::array<KV, 256> rather than
+      // a smaller-range vector with bounds, to keep locality and reduce
+      // comparisons. It didn't help.
+      std::vector<KV> next_contents;
+      next_contents.reserve(range_end - range_start);
+      std::copy(range_start, range_end, std::back_inserter(next_contents));
```

Comment:
Is `Value` expected to be a POD? Or might the value-copying be expensive. I wonder if KV should use a ptr or ref to Value.

Reply:
Oh, yeah, this should have been std::move (it was originally but it got lost in one of the various refactors). Making it so. (Otherwise it doesn't work with std::unique_ptr values - though that's fine in the single use-case there is right now, where it's just bare pointers.)